SLM Interval based scheduling

parkertimmins opened the pull request:
Add the ability to schedule an SLM policy with a time unit interval schedule rather than a cron job schedule. This is defined by using a a `schedule` argument in [time unit syntax](https://www.elastic.co/guide/en/elasticsearch/reference/current/api-conventions.html#time-units). For example:
```
PUT /_slm/policy/half-hour-snapshots
{
  "schedule": "30m",
  "name": "<half-hour-snap-{now/h}>",
  "repository": "repo",
  "retention": {
    "expire_after": "30d",
    "min_count": 5,
    "max_count": 50
  }
}
```

This policy will first run 30 minutes after the policy `modification_date`. It will then run again every time another 30 minutes has passed. Every time the policy is changed, the next snapshot will be re-scheduled to run one interval after the new modification date.

parkertimmins committed:
Add interval to SLM policy

- uses modifiedDate as interval_start_time
- dont run a snapshot immediately

parkertimmins committed:
fix some existing tests

parkertimmins committed:
fix some existing tests

parkertimmins committed:
spotless

parkertimmins committed:
cleanup last weeks work

parkertimmins committed:
fixes more existing tests

parkertimmins committed:
Add some unit tests for interval, inject clock

parkertimmins commented:
/ci

parkertimmins committed:
Merge branch 'main' into slm-interval-scheduling

parkertimmins committed:
Add more unit tests

parkertimmins committed:
Merge branch 'main' into slm-interval-scheduling

parkertimmins committed:
Add helper for testing schedule vs interval

parkertimmins committed:
add some docs

parkertimmins committed:
spotless

parkertimmins committed:
validation message change as schedule not required

parkertimmins commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/slm/SnapshotLifecyclePolicy.java:
```diff
@@ -287,6 +339,9 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
         builder.field(SCHEDULE.getPreferredName(), this.schedule);
+        if (this.interval != null) {
+            builder.field(INTERVAL.getPreferredName(), this.interval);
+        }
```
As it's currently written, if schedule is defined a GET request would not contain an interval field:
```
{
 "schedule": "* * * * * *",
...
}
```
whereas, if interval is defined, there would be a null (or maybe "null"?) schedule field:
```
{
 "schedule": null,
 "interval": "30m"
...
}
```
I dislike this asymmetry, but I prefer to not write out empty values. Unfortunately, we can't (easily) remove the schedule field, since this would be a breaking change.  The alternative would be to include a field with a null value in both cases. What do you think?

joegallo replied:
We should cover that scenario with other breaking changes policy experts -- it's only a breaking change **if you use the new feature** and I'm not sure we consider those breaking changes automatically.

parkertimmins replied:
Ah yeah I see what you mean. I think that makes sense that it would not be a breaking change in that case. I'll follow-up withe some bcc experts

joegallo replied:
I'm resolving this part of the conversation because we're no longer tracking an `interval` under its own key.

elasticsearchmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

elasticsearchmachine commented:
Hi @parkertimmins, I've created a changelog YAML for you.

parkertimmins committed:
Update docs/changelog/110847.yaml

joegallo commented on x-pack/plugin/slm/src/test/java/org/elasticsearch/xpack/slm/action/ReservedSnapshotLifecycleStateServiceTests.java:
```diff
@@ -111,9 +111,10 @@ public void testValidationFails() {
               }
             }""";
 
+        expectThrows(IllegalArgumentException.class, () -> processJSON(action, prevState, badPolicyJSON)).getMessage();
```
I'm not following the importance of this `expectThrows` versus the one after this inside the `assertThat` which looks identical to me.

parkertimmins replied:
Whoops! Yeah, that was me messing to understand what was happening with the following test case and shouldn't still be there!

joegallo commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/slm/SnapshotLifecyclePolicy.java:
```diff
@@ -81,14 +88,16 @@ public class SnapshotLifecyclePolicy implements SimpleDiffable<SnapshotLifecycle
         this.name = Objects.requireNonNull(name, "policy snapshot name is required");
-        this.schedule = Objects.requireNonNull(schedule, "policy schedule is required");
+        this.schedule = schedule;
+        this.interval = interval;
```
I would like this a bit better if there were validation in the constructor here that makes sure these cannot both be `null`. The previous logic required at this point in the constructor that the `schedule` wasn't `null`, and the validation that it was not an empty String was deferred to the `validate` method. I'd like to see something along the same lines, but handling both `interval` and `schedule`.

parkertimmins replied:
I want your opinion this my change for this. Probably easiest to look at the commit diff: https://github.com/elastic/elasticsearch/pull/110847/commits/1fa9058464fa4100c6471fbc9a4b4eae2e66fc24

I would like to validate that at least one of schedule and interval is defined, and that at most one of schedule and interval is defined. It turns out that the parser provides some helpers for this, along with nice error messages: 
```
PARSER.declareExclusiveFieldSet(SCHEDULE.getPreferredName(), INTERVAL.getPreferredName());
PARSER.declareRequiredFieldSet(SCHEDULE.getPreferredName(), INTERVAL.getPreferredName());
```

Unfortunately, these interact rather trickily with validation in the constructor. 
The assertion in the constructor:
```
assert schedule != null || interval != null : "policy schedule or interval is required";
```
does the same thing as `declareRequiredFieldSet`. It's still nice to have the assertion in case someone calls the constructor directly. This works fine.

But if we add a similar assertion for the `at most one is defined case`:
```
assert schedule == null || interval == null : "only one of schedule or interval may be defined";
```
The assertion bubbles up to the caller. Not the error message provided by the parser. It seems that the constructor is called eagerly during parsing. For this reason I haven't included this assertion in the constructor, because I'd like to use the error message from the parser. 

Anyway, I'm just not super happy with the inconsistency between the two cases. If you have any ideas, I'm all ears.

parkertimmins committed:
Add an example interval usage to docs

parkertimmins committed:
Add and fix tests

- add a test that checks interval against every minute of the day
- fix test which fails because random modification time fails validation due to being before current time

parkertimmins committed:
fail fast is both schedule and interval are null

parkertimmins committed:
Combine schedule and interval in single field

parkertimmins committed:
Some missing test fixes

joegallo commented:
Can you run a stock Kibana against a cluster running this code and see how it reacts to a `schedule` that it doesn't understand? That is, what happens if you manually put in an SLM policy that uses an interval rather than a cron expression and then try to load it up in the Kibana SLM UI? For that matter, will Kibana allow you to create new policies using such syntax or would this PR require some work on Kibana in order to take advantage of the new feature?

parkertimmins commented:
Good news! I think it just requires some description updates. There is a schedule input UI when you can access the page:

![image](https://github.com/user-attachments/assets/cd73a4dd-21b6-45b6-81a2-02391bd5a4b8)
But if you click "Create cron expression" you get to the manual input. **We'll need to update this to "Create cron or time unit expression"**
![image](https://github.com/user-attachments/assets/b74cc028-4fb7-4570-a501-3286cab9c746)

This input allows an interval expression. I'm not sure if the red highlighting is due to actual cron validation or that fact that it's not a word. **We should change this highlighting or probably just turn it off on the element. Also we'll need to change the "Use cron expression" and link.**
![image](https://github.com/user-attachments/assets/afd2dfa4-3370-4e80-992e-ea2054eff5ec)

Submitting this successfully creates policies with interval schedules. (I've reduced the minimum interval setting to 1s for testing).
![image](https://github.com/user-attachments/assets/906b9ab1-df33-4311-9f4b-df5db6768101)

I've also confirmed that a policy with an interval schedule can be re-opened for editing.

parkertimmins committed:
Small doc and test update

joegallo committed:
Merge branch 'main' into slm-interval-scheduling

joegallo committed:
Drop unused transport version

masseyke committed:
Fixing potential OOME in HistoryIntegrationTest.testThatHistoryIsTruncated() (#111556)

joegallo committed:
Add a comment

joegallo commented on server/src/main/java/org/elasticsearch/common/scheduler/SchedulerEngine.java:
```diff
@@ -42,10 +42,16 @@ public class SchedulerEngine {
     public static class Job {
```
Can `Job` and `Event` (and maybe some of the other things here) be proper `Record`s? If so, I'm indifferent to whether that's done on this PR or a subsequent PR, but I would like to see it done if it's possible to do.

Separately, I think `startTime` needs an `@Nullable` annotation and some commentary on **why** somebody should or should not set it. It worries me that there's this extra option with no manual.

joegallo commented:
Can you add an integration test that verifies that the `next_execution_millis` that comes back from a `GET _slm/policy` call for an interval-based policy schedule is correct? I imagine it will be, but this is a place where the new logic would get exercised end-to-end.

joegallo commented:
Off the top of my head, I'm not sure about how the serialization is going to work here -- it's been a little while since I've been certain of where SLM runs and what our defined upgrade process is (there's a lot of dust on those neurons):

If I have an 8.15.2 cluster running with one master node and two data nodes, what happens when I'm in a partially upgraded state such that at least one node in my cluster is able to receive an interval based schedule but there are still other nodes in my cluster that can't understand them?

IIRC SLM runs on the master, and IIRC we require that master-eligible nodes are upgraded first, so maybe everything would be just fine...

alternatively, maybe you need some other way of handling the serialization so that we don't just immediately start throwing exceptions on the old nodes when they receive a new object and can't parse the schedule as a cron expression (since their code only has the ability to parse as cron expressions!).

joegallo commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/slm/SnapshotLifecyclePolicy.java:
```diff
@@ -129,23 +130,45 @@ public SnapshotRetentionConfiguration getRetentionPolicy() {
+    /**
+     * @return whether `schedule` is a cron expression, as opposed to a time unit
+     */
+    public boolean isCronSchedule() {
```
Ehhhhhhhhh, running a try/catch here each time somebody invokes this kinda rubs me the wrong way. I would feel a bit better about it if it were executed once in the constructor and remembered forever.

joegallo commented:
I've added some comments, but since I'm going to be out next week I added @dakrone as a reviewer to continue the conversation in my place (feel free to delegate to somebody else, though!).

parkertimmins committed:
Register SLM run before snapshotting to save stats (#110216)

The SLM health indicator relies on the policyMetadata.getInvocationsSinceLastSuccess to determine if the last several snapshots have failed. If a snapshot fails and the master is shutdown before setting invocationsSinceLastSuccess, the fact that failure occurred will be lost.

To solve this, before snapshotting, we register that a snapshot is about to run, in the cluster state custom metadata. If the run fails, and invocationsSinceLastSuccess is not updated before a master shutdown, the fact that the failure occurred will not be lost. On completion of a subsequent snapshot run, SnapshotLifecycleTask will observe that there exists a registered snapshot which is no longer running. It will infer that the snapshot failed, and update invocationsSinceLastSuccess and other stats accordingly.

A few parts of this change touch general snapshot code, and are worth noting:
* Snapshots can only be uniquely identified with a uuid which had previously been generated in SnapshotService. This uuid is needed when there is a snapshot failure, but was not available to SLM as it was only returned in the SnapshotInfo after a success. To make this available, the uuid is now generated in the CreateSnapshotRequest constructor and passed to SnapshotService. In mixed version clusters, there exists a special case where the uuid is still generated in SnapshotService.
* If a snapshot were registered before calling snapshot service, there would a small period of time when a snapshot is registered but not yet stored in SnapshotsInProgress. During this time, another snapshot from the same policy might incorrectly infer that the snapshot failed and update its stats accordingly. To avoid this race condition, we register snapshots within SnapshotService in the same cluster update which updates SnapshotsInProgress.

parkertimmins committed:
Cleanup - change scheduler job/event to records

parkertimmins committed:
add comment and refactor how schedule built

parkertimmins committed:
Add IT test of next_execution_millis

parkertimmins committed:
Add slm.interval_schedule node feature

parkertimmins committed:
Merge branch 'main' into slm-interval-scheduling

parkertimmins committed:
just a bit of cleanup

parkertimmins committed:
Rand interval more than 15m to fix IT test

parkertimmins committed:
spotless

joegallo committed:
Merge branch 'main' into pull/110847

joegallo committed:
Nit: make these the same schedule

joegallo reviewed (approved):
🚀, this a very cool feature.

parkertimmins commented:
@elasticsearchmachine  run elasticsearch-ci/part-3
